goto-coverage bug fix

ChenfengWei0 opened the pull request:
ESBMC shows incorrect coverage information when the program contains loop/func_call

- command: `esbmc main.c --make-assert-false --multi-property`


example1
```c++
void loop1()
{
   for (int i = 0; i < 2; i++)
   {
      assert(1);
   }
}

void loop2()
{
   for (int i = 0; i < 2; i++)
   {
      assert(1);
   }
}

int main()
{

   switch (nondet_int())
   {
   case 1:
      loop1();
   case 2:
      loop2();
   default:
      ;
   }
   return 0;
}
```
Before fixing:
<img width="164" alt="df45726bcc7e17fe34abb0778a85cd5" src="https://github.com/esbmc/esbmc/assets/105888669/4a38a160-2308-4f92-a291-fccb51a97b9a">

After:
<img width="167" alt="532f338deb0b61a5d393c7e3bea919c" src="https://github.com/esbmc/esbmc/assets/105888669/4d68dd82-3d3f-4363-8192-d6b440a3490f">

example2
```c++
void t1()
{
   assert(1==2);
}

void t2()
{
   assert(1==2);
}

int main()
{
   t1();
   t2();
   t1();
   return 0;
}
```
Before fixing:
<img width="174" alt="11625fe49a746df9fa40cc10c8f3f65" src="https://github.com/esbmc/esbmc/assets/105888669/4b9d9e8b-936f-4b2c-b7b3-ebfb4d285e2a">

After:
<img width="167" alt="532f338deb0b61a5d393c7e3bea919c" src="https://github.com/esbmc/esbmc/assets/105888669/ea0b4ddf-f0e5-42d7-a92a-9f7c3a1b26d5">

ChenfengWei1 committed:
goto-coverage bug fix

ChenfengWei1 committed:
lint code base

ChenfengWei1 committed:
lint code base

mikhailramalho commented:
why do we count the assertion only once when it's reachable multiple times inside a loop?

ChenfengWei0 commented:
> why do we count the assertion only once when it's reachable multiple times inside a loop?

The original idea is to check if the piece of code (branch/function/loop) is reachable. This is done by checking if the assertion inside is reachable.

ChenfengWei1 committed:
lint code base

mikhailramalho reviewed (changes requested):
We should avoid throwing exception, you can call log_error and abort instead.

I also see a lot of duplicated code with other methods that do the bmc, stuff like creating the solver, running symex/generating the equation, and printing the cex, can we unify these calls?

ChenfengWei0 commented:
> We should avoid throwing exception, you can call log_error and abort instead.
> 
> I also see a lot of duplicated code with other methods that do the bmc, stuff like creating the solver, running symex/generating the equation, and printing the cex, can we unify these calls?

Yes, I agree the multi-property code is messy. I will have another PR to refactor the code.

fbrausse commented on src/esbmc/bmc.cpp:
```diff
@@ -731,104 +732,118 @@ smt_convt::resultt bmct::multi_property_check(
+            {
+              // avoid double counting the claims in loop/func_call
               tracked_instrument++;
+              reached_claims.insert(claim.claim_msg);
```
This pattern of a set.find() (or set.count()) condition followed by set.insert() is inefficient. Suggest to use `if(auto [it, ins] = set.emplace(...); ins)` instead.

ChenfengWei0 replied:
Changed as suggested

fbrausse commented on src/esbmc/bmc.cpp:
```diff
@@ -713,6 +713,7 @@ smt_convt::resultt bmct::multi_property_check(
   std::mutex result_mutex;
   // For coverage info
   int tracked_instrument = 0;
+  std::unordered_set<std::string> reached_claims;
```
Two questions:
- It seems `tracked_instrument` is the same as `reached_claims.size()`. Do we need both?
- Is a claim indeed uniquely identified by its `claim_msg`, which seems to be a somewhat static comment? If so, could you explain how that works?

ChenfengWei0 replied:
1. Deleted  ‘tracked_instrument’
2. Explained as below

Thanks for the suggestions.

fbrausse commented:
> > why do we count the assertion only once when it's reachable multiple times inside a loop?
> 
> The original idea is to check if the piece of code (branch/function/loop) is reachable. This is done by checking if the assertion inside is reachable.

Could we avoid generating duplicate results instead of massaging the quantities output to the user? I imagine a higher iteration count in the source program would result in a lot of wasted resources.

ChenfengWei0 commented:
> > > why do we count the assertion only once when it's reachable multiple times inside a loop?
> > 
> > 
> > The original idea is to check if the piece of code (branch/function/loop) is reachable. This is done by checking if the assertion inside is reachable.
> 
> Could we avoid generating duplicate results instead of massaging the quantities output to the user? I imagine a higher iteration count in the source program would result in a lot of wasted resources.

Thank you for the suggestions. Will check

lucasccordeiro commented:
Thanks, @ChenfengWei0.

Please let us know once you have a revised version of this PR.

mikhailramalho commented:
maybe we could show how many times an assertion was reached by esbmc?

Em qua., 13 de set. de 2023 às 04:50, Franz Brauße ***@***.***>
escreveu:

> why do we count the assertion only once when it's reachable multiple times
> inside a loop?
>
> The original idea is to check if the piece of code (branch/function/loop)
> is reachable. This is done by checking if the assertion inside is reachable.
>
> Could we avoid generating duplicate results instead of massaging the
> quantities output to the user? I imagine a higher iteration count in the
> source program would result in a lot of wasted resources.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/1337#issuecomment-1717122438>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH4ENUCRFTB5OZHFNKTX2FQTNANCNFSM6AAAAAA4VNLD2E>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

Mikhail R. Gadelha, Ph.D.

ChenfengWei0 commented:
> maybe we could show how many times an assertion was reached by esbmc? Em qua., 13 de set. de 2023 às 04:50, Franz Brauße ***@***.***> escreveu:
> […](#)

Yes it is doable, but I'm not sure of the need or purpose of this feature. Should it be a new option to to enable to show this information?

fbrausse commented:
My understanding: If ESBMC is meant to produce this kind of coverage info based on asserts, there should be a single cmdline switch for that. The current combination `--make-assert-false --multi-property` is just an approximation. The new cmdline switch would make sure to not repeatedly try to determine whether the same assert is reachable.

Obviously the current PR isn't addressing this, I just thought it could be a useful approach. But I'm not familiar with the coverage usage of ESBMC, so I just put this comment here in case someone familiar with it can comment on whether that's feasible.

mikhailramalho commented:
BTW, why do we need to make all assertions false to check reachability?

We can easily check if an instruction is reachable by checking if its guard
is true.

On Wed, 13 Sept 2023, 12:35 Franz Brauße, ***@***.***> wrote:

> My understanding: If ESBMC is meant to produce this kind of coverage info
> based on asserts, there should be a single cmdline switch for that. The
> current combination --make-assert-false --multi-property is just an
> approximation. The new cmdline switch would make sure to not repeatedly try
> to determine whether the same assert is reachable.
>
> Obviously the current PR isn't addressing this, I just thought it could be
> a useful approach. But I'm not familiar with the coverage usage of ESBMC,
> so I just put this comment here in case someone familiar with it can
> comment on whether that's feasible.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/1337#issuecomment-1717870167>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH6JJGKPPRVAM3XQF43X2HHEHANCNFSM6AAAAAA4VNLD2E>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

lucasccordeiro commented:
@ChenfengWei0: can I ask you for the status of this issue?

ChenfengWei0 commented:
This PR has fixed the bugs goto-coverage with loops. 

But there are some other bugs still unfixed:
- https://github.com/esbmc/esbmc/issues/1361
- https://github.com/esbmc/esbmc/issues/1399

I plan to fix these all together in one PR. And the refactoring of the multi-property can be another PR.

ChenfengWei1 committed:
fix double counting/verifying in multi-property

ChenfengWei1 committed:
redesign goto-coverage option

ChenfengWei1 committed:
avoid removal in multi-property

ChenfengWei1 committed:
add option goto-coverage-claims

ChenfengWei1 committed:
lint code base

ChenfengWei1 committed:
fix typo

ChenfengWei1 committed:
show warnings

ChenfengWei1 committed:
bug fix

ChenfengWei0 commented:
Update:
1. avoid double verifying/counting the asserts in loop unwinding/func call (mentioned above). We will report "skipping" when we meet a verified claim.

![image](https://github.com/esbmc/esbmc/assets/105888669/3fd71567-546a-4bd2-b166-cf3c6bf08968)

2. If the user wants to verify all claims, we now have an option `keep-verified-claims`.

![image](https://github.com/esbmc/esbmc/assets/105888669/53a5edab-2a42-4a6f-979e-bd59c7bc984e)


3. redesign `goto-coverage` option and add a new option `goto-coverage-claims`

```
  --goto-coverage                       this activates --make-assert-false and
                                        --multi-property and deactivates
                                        --keep-unwind-claims
  --goto-coverage-claims                enable goto-coverage and shows all
                                        reached claims
```

`goto-coverage`
![image](https://github.com/esbmc/esbmc/assets/105888669/8cd12c6a-57df-4b83-a6e5-25b1f1522d6c)

`goto-coverage-claims`

![image](https://github.com/esbmc/esbmc/assets/105888669/535ff54e-8896-4c5b-af39-0e160ddf189c)

ChenfengWei1 committed:
fix typo

ChenfengWei1 committed:
show unreached claims

ChenfengWei0 commented:
Update:
4. Show unreached claims:

e.g. (`--goto-coverage-claims --k-induction --color`)
```c++
int main()
{
   int x = 0;
   while (nondet_int())
   {
      if (!x)
      {
         assert(x == 0);
         x = 1;
      }
      else if (x == 1)
      {
         assert(x > 0);
         x = 2;
      }
      else if (x == 2)
      {
         assert(x >= 2);
         x = 3;
      }
   }
   assert(x == 3);
}
```

**Result**

![image](https://github.com/esbmc/esbmc/assets/105888669/b9704074-e375-4628-b661-c1ea4a78a9f3)

lucasccordeiro commented:
comments from the ESBMC's user:

````
For some reason I’m unable to login with my account to GitHub, so I’m writing my comments here.

 

I think that it is important to know about multiple instances of an assertions and whether they were covered. So, using Chenfeng’ example:

 

void loop1()

{

   for (int i = 0; i < 2; i++)

   {

      assert(1);

   }

}

 

void loop2()

{

   for (int i = 0; i < 2; i++)

   {

      assert(1);

   }

}

 

int main()

{

 

   switch (nondet_int())

   {

   case 1:

      loop1();

   case 2:

      loop2();

   default:

      ;

   }

   return 0;

}

 

The report should be:

Total assets: 2

Total assertions instances: 4

Reached assertions instances:

Coverage (with respect to assertions instances!)
````

ChenfengWei1 committed:
count assert instance

ChenfengWei0 commented:
Update:

## exmaple1
```c++
void loop1()
{
   for (int i = 0; i < 2; i++)
   {
      assert(1);
   }
}

void loop2()
{
   for (int i = 0; i < 2; i++)
   {
      assert(1);
   }
}

int main()

{
   switch (nondet_int())
   {
   case 1:
      loop1();
   case 2:
      loop2();
   default:
       ;
   }
   return 0;
}
```
`--unwind 2 --no-unwinding-assertions --goto-coverage-claims --color`

![image](https://github.com/esbmc/esbmc/assets/105888669/39d482d9-4350-46e5-ae64-41a981e906a0)

## exmaple2

```c++
// loop1, loop2 same as example1
int main()

{
   switch (1)
   {
   case 1:
      loop1();
   case 2:
      loop2();
   default:
       ;
   }
   return 0;
}
```

![image](https://github.com/esbmc/esbmc/assets/105888669/5e5c4928-1306-4964-83ab-5035f0b80809)


## example3
```c++
int main()
{
   int x = 0;
   while (nondet_int())
   {
      if (!x)
      {
         assert(x == 0);
         x = 1;
      }
      else if (x == 1)
      {
         assert(x > 0);
         x = 2;
      }
      else if (x == 2)
      {
         assert(x >= 2);
         x = 3;
      }
   }
   assert(x == 3);
}
```
`--k-induction --goto-coverage-claims --color --show-claims`

![image](https://github.com/esbmc/esbmc/assets/105888669/157eae9b-b983-491e-9a16-191829d8b4a4)


- **Note:** We can always check unreached asserts by comparing the output of `--goto-coverage-claims` and `--show-claims`

![image](https://github.com/esbmc/esbmc/assets/105888669/6ee373f9-d22d-4cb1-aa6f-7e7f7ed268ee)

ChenfengWei1 committed:
fix claim number

fbrausse commented on src/esbmc/bmc.cpp:
```diff
@@ -731,74 +735,102 @@ smt_convt::resultt bmct::multi_property_check(
+          !reached_claims.count(loc) ||
+          options.get_bool_option("keep-verified-claims"))
+        {
+          reached_claims.insert(loc);
```
Again: `insert` following `count`/`find` is inefficient; see https://github.com/esbmc/esbmc/pull/1337#discussion_r1324084773

ChenfengWei0 replied:
Will change

fbrausse commented on src/esbmc/bmc.cpp:
```diff
@@ -711,8 +711,12 @@ smt_convt::resultt bmct::multi_property_check(
   std::mutex result_mutex;
+  // Store the location of the assertion
+  // to avoid double verifying the unwinding claims in loop/func_call
+  std::string loc = "";
```
Why does this variable `loc` need to live outside of the job_function? It only seems to be used in the `if(result == smt_convt::P_SATISFIABLE)` block. Can't you move it inside?

ChenfengWei0 replied:
My mistake

fbrausse commented on src/esbmc/bmc.cpp:
```diff
@@ -731,74 +735,102 @@ smt_convt::resultt bmct::multi_property_check(
+            // which has the same location (e.g. add-false-assert)
+            // this can also be used in goto-cov output
+            // we use the location (and comment) to distinguish each claim
+            loc = step.comment + "\t" + loc;
```
Could you please explain why the step.pc->location.as_string() combined with the step.comment is unique?

ChenfengWei0 replied:
- Why `comment`: In goto-coverage, we add a claim number prefix in the comment. e.g. `Claim 1: ...`. 
- Why `location`: We should also consider the unwinding assertions, which are not processed by goto-cov. e.g.
```
[Coverage]

Total Asserts   : 2
Total Assertion Instances: 6
Reached Assertions Instances:
  unwinding assertion loop      file 2.c line 1248 column 4 function loop1
  Claim 2: assertion 1  file 2.c line 1258 column 7 function loop2
  Claim 2: assertion 1  file 2.c line 1258 column 7 function loop2
  Claim 1: assertion 1  file 2.c line 1250 column 7 function loop1
  Claim 1: assertion 1  file 2.c line 1250 column 7 function loop1
  unwinding assertion loop      file 2.c line 1256 column 4 function loop2
Coverage: 100%
```
- There could be asserts that have the same location (e.g. add-false-assert)

Therefore, we need both comment and location.

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1648,10 +1648,16 @@ bool esbmc_parseoptionst::process_goto_program(
 
     // Start by removing all no-op instructions and unreachable code
-    if(!cmdline.isset("no-remove-no-op"))
+    // We should skip this removal in goto-cov and multi-property
```
Why? Can you please extend the comment with an explanation?

ChenfengWei0 replied:
Because multi-property wants to find all the bugs in the src code. But these optimisations will remove the rest of the code during the Goto stage. e.g.

```c++
int main()

{
   int x[2];
   x[3] = 10;
   switch (nondet_int())
   {
   case 1:
      assert(0);
   case 2:
      assert(0);
   default:
       ;
   }
   return 0;
}
```

will only show one counterexample in multi-property mode:
```
[Counterexample]


State 1 file 2.c line 1251 column 4 function main thread 0
----------------------------------------------------
Violated property:
  file 2.c line 1251 column 4 function main
  array bounds violated: array `x' upper bound
  0


VERIFICATION FAILED
```

ChenfengWei0 replied:
Also, the idea of goto-coverage is to find out unreached codes(i.e. asserts). Thus we cannot remove the unreachable code beforehand.

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1648,10 +1648,16 @@ bool esbmc_parseoptionst::process_goto_program(
+    // We should skip this removal in goto-cov and multi-property
+    if(!(cmdline.isset("no-remove-no-op") || cmdline.isset("multi-property") ||
+         cmdline.isset("parallel-solving") || cmdline.isset("goto-coverage") ||
+         cmdline.isset("goto-coverage-claims")))
```
As the set of options here is quite extensive and the condition being checked repeatedly, it makes sense to store it in a boolean variable with an appropriate name.

ChenfengWei0 replied:
Changed as suggested.

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1648,10 +1648,16 @@ bool esbmc_parseoptionst::process_goto_program(
+         cmdline.isset("multi-property") || cmdline.isset("parallel-solving") ||
+         cmdline.isset("goto-coverage") ||
+         cmdline.isset("goto-coverage-claims")))
       remove_unreachable(goto_functions);
```
Why do any of these newly added options prohibit removing unreachable code?

ChenfengWei0 replied:
Explained as above.

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1648,10 +1648,16 @@ bool esbmc_parseoptionst::process_goto_program(
+    if(!(cmdline.isset("no-remove-no-op") || cmdline.isset("multi-property") ||
+         cmdline.isset("parallel-solving") || cmdline.isset("goto-coverage") ||
+         cmdline.isset("goto-coverage-claims")))
       remove_no_op(goto_functions);
```
Why do any of these newly added options prohibit removing no-ops?

ChenfengWei0 replied:
Removed.

Both Multi-property and goto-coverage only need to enable ‘no—remove-unreachable’.

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1702,10 +1708,15 @@ bool esbmc_parseoptionst::process_goto_program(
+    if(!(cmdline.isset("no-remove-no-op") || cmdline.isset("multi-property") ||
+         cmdline.isset("parallel-solving") || cmdline.isset("goto-coverage") ||
+         cmdline.isset("goto-coverage-claims")))
       remove_no_op(goto_functions);
```
Same as above.

ChenfengWei0 replied:
Explained as above.

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1702,10 +1708,15 @@ bool esbmc_parseoptionst::process_goto_program(
+         cmdline.isset("multi-property") || cmdline.isset("parallel-solving") ||
+         cmdline.isset("goto-coverage") ||
+         cmdline.isset("goto-coverage-claims")))
       remove_unreachable(goto_functions);
```
Same as above.

ChenfengWei0 replied:
Explained as above.

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1721,17 +1732,32 @@ bool esbmc_parseoptionst::process_goto_program(
+
+    if(
+      options.get_bool_option("make-assert-false") ||
+      cmdline.isset("make-assert-false"))
```
Why do we need to check both, the `options` and the `cmdline`?

ChenfengWei0 replied:
my mistake

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1721,17 +1732,32 @@ bool esbmc_parseoptionst::process_goto_program(
+      // for assertion coverage metric
+      options.set_option("make-assert-false", true);
+      // no-simplify, otherwise the coverage will always be 100%
+      options.set_option("no-simplify", true);
```
> `no-simplify, otherwise the coverage will always be 100%`

I don't understand. Could you explain why that is the case?

ChenfengWei0 replied:
My investigation is:
- if disables `no-simplify`, the coverage( reached_assert_instance / total_assertion_instance) will always be 100%. 
- It seems that, after simplifying, some unreachable(?) claim instances are missing in `result->remaining_claims` (from `multi_property_check(eq, result->remaining_claims)`). This leads to the 100% coverage as all the remaining_claims are reachable claims.

ChenfengWei1 committed:
fix issue #1408

ChenfengWei0 commented:
Fix issue: #1408

fbrausse reviewed (changes requested):
I have a couple of comments below. Some are requests for clarifications because this code already is messy and this PR is not helping with that. It's also unclear how all these new options get combined.

Could you please put a comment at the beginning of the multi_property_check() mentioning each option being used in this function directly and what purpose it serves?

fbrausse commented:
> Fix issue: #1408

@ChenfengWei0 sorry, that's not the right fix. It just avoids generating the counter-example while the bug is inside the generation itself (arrays as tuple elements not supported by `smt_convt::get()` - or rather the tuple_node_flattener).

ChenfengWei0 commented:
> > Fix issue: #1408
> 
> @ChenfengWei0 sorry, that's not the right fix. It just avoids generating the counter-example while the bug is inside the generation itself (arrays as tuple elements not supported by `smt_convt::get()` - or rather the tuple_node_flattener).

Yes. Maybe it should be "avoid triggering this bug" instead of "bug fix". Still, I think we need this commit even if there is no such bug.

fbrausse commented:
> Still, I think we need this commit even if there is no such bug.

Not sure, maybe. Any reasons?

Independently: This PR is going to need a few regression tests that check the new behaviour.

ChenfengWei0 commented:
> > Still, I think we need this commit even if there is no such bug.
> 
> Not sure, maybe. Any reasons?
> 
> Independently: This PR is going to need a few regression tests that check the new behaviour.

I just thought we should keep the same process of building error trace as the function `void bmct::error_trace()`, which is used in non-multi-property mode. (That's also the reason why issue #1408 only discovered in multi-property).

i.e.
```c++
  bool is_compact_trace = true;
  if(
    options.get_bool_option("no-slice") &&
    !options.get_bool_option("compact-trace"))
    is_compact_trace = false;

  goto_tracet goto_trace;
  build_goto_trace(eq, smt_conv, goto_trace, is_compact_trace);
```

Also, users may want to use `no-slice` and `compact-trace` in multi-property.

---

Will add regression tests

ChenfengWei1 committed:
remove parallel-solving

ChenfengWei1 committed:
refactoring reached_claims

ChenfengWei1 committed:
add  boolean variable for mul and cov

ChenfengWei1 committed:
fix typo

ChenfengWei1 committed:
fix typo

ChenfengWei1 committed:
remove unnecessary codes

ChenfengWei1 committed:
fix codacy

ChenfengWei1 committed:
bug fix

ChenfengWei1 committed:
add regression tests

ChenfengWei1 committed:
change regression test

ChenfengWei1 committed:
change regression test

ChenfengWei1 committed:
fix typo

ChenfengWei1 committed:
change regression test

ChenfengWei1 committed:
fix regression test

ChenfengWei1 committed:
fix regression test

lucasccordeiro reviewed (approved):
Thanks for submitting this PR, @ChenfengWei0.

lucasccordeiro commented:
@ChenfengWei0: Can I ask whether you have addressed all comments from @fbrausse? If they don't make sense, please explain so @fbrausse can interact with you.

ChenfengWei0 commented:
> @ChenfengWei0: Can I ask whether you have addressed all comments from @fbrausse? If they don't make sense, please explain so @fbrausse can interact with you.

I believe I have addressed all the comments. In addition, I have removed the ‘parallel-solving’ options according to the discussions in #1028 .

lucasccordeiro commented:
Many thanks, @ChenfengWei0.

@fbrausse: Can I ask you whether you still have further comments?

lucasccordeiro commented:
Thanks for submitting this PR, @ChenfengWei0.

fbrausse commented on src/esbmc/bmc.cpp:
```diff
@@ -726,79 +730,104 @@ smt_convt::resultt bmct::multi_property_check(
+              loc = step.pc->location.as_string();
+
+            // we use the "comment + location" to distinguish each claim
+            // - locaiton: ideally, the loc is enough for distinguishuing claims.
```
Typos: location and distinguishing. Also: Why "ideally"? Is it or it is not enough to distinguish claims? I would actually be surprised if it was because we butcher the location somewhat in <https://github.com/esbmc/esbmc/blob/cc230ed73c5f7b51ec6381295ce07e151649fbc2/src/clang-c-frontend/clang_c_convert.cpp#L3421> and in <https://github.com/esbmc/esbmc/blob/cc230ed73c5f7b51ec6381295ce07e151649fbc2/src/clang-c-frontend/clang_c_convert.cpp#L3418>.

ChenfengWei0 replied:
My assumption is that the location of each claim is unique ideally, as they have different line numbers. But some mutations/checkings in ESBMC (e.g. add-false-assert) will add the instructions with the same location as the previous instruction. 

Still, I think the comment is not clear enough and not necessary. The core idea should be:
- comment + location : `Claim x: ... location line y`
- x is unique for sure. However, the unwinding asserts do not have these `Claim x` prefixes. Therefore we add the location behind, as the line number for each unwinding assert is different.

Will Fix typo.

fbrausse replied:
Right, that makes more sense, thanks!

fbrausse commented on src/esbmc/bmc.cpp:
```diff
@@ -810,68 +839,56 @@ smt_convt::resultt bmct::multi_property_check(
+    if(total_instance != 0)
+      log_result("Coverage: {}%", tracked_instance * 100.0 / total_instance);
+    else
+      log_result("Coverage: 0%");
```
We may want to relabel this. "Coverage" usually has a quite specific meaning, which differs from what is being tested here by replacing asserts by assert(0) and testing each one separately. @mikhailramalho's comment https://github.com/esbmc/esbmc/pull/1337#issuecomment-1717900619 is very relevant to the original meaning of coverage:

>  We can easily check if an instruction is reachable by checking if its guard
is true.

ChenfengWei0 replied:
I agree. It is no longer the original meaning of coverage.

Maybe we should change it to something like `Assertion Instance Coverage`, which is reached_instance/total_instance. The total assertion instances include the assert inside the source file, the unwinding asserts, the claims inserted during the goto-check and so on.

fbrausse replied:
I agree, with this explanation also given in the help message for the `--goto-coverage` option that would make it more clear.

fbrausse reviewed (commented):
Minor comments, otherwise looks good. Still I think branding this algorithm as producing "coverage information" is misleading.
